sdk-cpp/src/internal/com/download_impl.cpp

```diff
@@ -186,109 +209,134 @@ class DOStatusCallback :
+    // Need to provide this definition because try/catch keywords in the implementation above will fail builds with exceptions disabled
+    IFACEMETHODIMP OnStatusChange(IDODownload* download, const DO_DOWNLOAD_STATUS* comStatus)
+    {
+        msdo::download_status status = ConvertFromComStatus(*comStatus);
+        _callback(*_download, status);
+        return S_OK;
+    }
+
+#endif
 
 private:
     msdo::download_property_value::status_callback_t _callback;
     msdo::download* _download;
 };
 
+CDownloadImpl::CDownloadImpl(const std::string& uri, const std::string& downloadFilePath) noexcept
+{
+    _codeInit = _Initialize(uri, downloadFilePath);
+}
 
-CDownloadImpl::CDownloadImpl(const std::string& uri, const std::string& downloadFilePath)
+int32_t CDownloadImpl::_Initialize(const std::string& uri, const std::string& downloadFilePath) noexcept
 {
     Microsoft::WRL::ComPtr<IDOManager> manager;
-    throw_if_fail(CoCreateInstance(__uuidof(DeliveryOptimization), nullptr, CLSCTX_LOCAL_SERVER, IID_PPV_ARGS(&manager)));
+    RETURN_RES_IF_FAILED(CoCreateInstance(__uuidof(DeliveryOptimization), nullptr, CLSCTX_LOCAL_SERVER, IID_PPV_ARGS(&manager)));
     Microsoft::WRL::ComPtr<IDODownload> spDownload;
-    throw_if_fail(manager->CreateDownload(&spDownload));
+    RETURN_RES_IF_FAILED(manager->CreateDownload(&spDownload));
 
-    try {
-        throw_if_fail(CoSetProxyBlanket(static_cast<IUnknown*>(spDownload.Get()), RPC_C_AUTHN_DEFAULT,
-            RPC_C_AUTHZ_NONE, COLE_DEFAULT_PRINCIPAL, RPC_C_AUTHN_LEVEL_DEFAULT, RPC_C_IMP_LEVEL_IMPERSONATE,
-            nullptr, EOAC_STATIC_CLOAKING));
+    RETURN_RES_IF_FAILED(CoSetProxyBlanket(static_cast<IUnknown*>(spDownload.Get()), RPC_C_AUTHN_DEFAULT,
+        RPC_C_AUTHZ_NONE, COLE_DEFAULT_PRINCIPAL, RPC_C_AUTHN_LEVEL_DEFAULT, RPC_C_IMP_LEVEL_IMPERSONATE,
+        nullptr, EOAC_STATIC_CLOAKING));
 
-        download_property_value propUri(uri);
-        download_property_value propDownloadFilePath(downloadFilePath);
+    download_property_value propUri(uri);
+    download_property_value propDownloadFilePath(downloadFilePath);
+
+    RETURN_RES_IF_FAILED(_SetPropertyHelper(*spDownload.Get(), download_property::uri, propUri));
+    RETURN_RES_IF_FAILED(_SetPropertyHelper(*spDownload.Get(), download_property::download_file_path, propDownloadFilePath));
 
-        _SetPropertyHelper(*spDownload.Get(), download_property::uri, propUri);
-        _SetPropertyHelper(*spDownload.Get(), download_property::download_file_path, propDownloadFilePath);
-    }
-    catch (const exception& e)
-    {
-        spDownload->Abort();
-        throw e;
-    }
     _spDownload = std::move(spDownload);
+    return static_cast<int32_t>(errc::s_ok);
 }
 
 // Support only full file downloads for now
-void CDownloadImpl::Start()
+int32_t CDownloadImpl::Start() noexcept
 {
+    RETURN_RES_IF_FAILED(_codeInit);
```

Comment:
>RETURN_RES_IF_FAILED(_codeInit); [](http://example.com/codeflow?start=4&length=32)

Yeah this type of thing is a code smell. I'll describe the cleaner option in the do_download_property.h comment.